File: tests/test_rpim_2.cc
/*
GoGoRPIM - A RPIM implementation and 2-D Eletromagnetic simulator
    Copyright (C) 2012  Péricles Lopes Machado (LANE-UFPA)

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/

/**Copyright 2012 Péricles Lopes Machado
@file test_rpim_2.cc

UFPA - Universidade Federal do Pará
Laboratório de Análises Numéricas em Eletromagnetismo (LANE) 

Implementação do RPIM - Radial Point Interpolation Method

Esta implementação faz parte do TCC de Engenharia da Computaçã de Péricles Lopes Machado (0608000101)
*/

#include <algorithm>

#include "test_rpim_2.h"

using namespace lane_maxwell::modules::matrix;
using namespace lane_maxwell::modules::point;
using namespace lane_maxwell::modules::rpim;
using namespace lane_maxwell::modules::domains;

Point* point3D(double x, double y, double z, double eps, double mu, int prop, int ID, char type) {
	Point* po = new Point;
	Vector* p = new Vector(3);

	p->set(0, x);
	p->set(1, y);
	p->set(2, z);

	po->set(ID, eps, mu, prop, type, p);
	return po;
}

void genPoints(Points* const vp, int nx, int ny, int nz, char type) {

	double dx = 0.1;
	double dy = 0.1;
	double dz = 0.1;

	double eps = 8.841941e-12;
	double mu = 1.256637e-6; 
	int prop = 0;
	static int ID = 0;
	//void set(int ID, double eps, double mu, int prop, char type,
				//matrix::Vector* const p)
	for (int ix = 0; ix < nx; ++ix) {
		for (int iy = 0; iy < ny; ++iy) { 
			for (int iz = 0; iz < nz; ++iz) {
				double x, y, z;
				if (type == 'E') {
					x = (2*ix+0)*dx;
					y = (2*iy+0)*dy;
					z = (2*iz+0)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+0)*dx;
					y = (2*iy+1)*dy;
					z = (2*iz+1)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+1)*dx;
					y = (2*iy+1)*dy;
					z = (2*iz+0)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+1)*dx;
					y = (2*iy+0)*dy;
					z = (2*iz+1)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;
				} else {
					x = (2*ix+1)*dx;
					y = (2*iy+1)*dy;
					z = (2*iz+1)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+1)*dx;
					y = (2*iy+0)*dy;
					z = (2*iz+0)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+0)*dx;
					y = (2*iy+0)*dy;
					z = (2*iz+1)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;

					x = (2*ix+0)*dx;
					y = (2*iy+1)*dy;
					z = (2*iz+0)*dz;
					vp->push_back(point3D(x, y, z, eps, mu, prop, ID, type));
					++ID;


				}
			}
		}
	}
}

double dumb(const Points* p, int T, char A, char B) {
	double acc = 0;

	for (int i = 0; i < p->size(); ++i) {
		if(p->at(i)->getType() == A) {
			std::vector<double> d;

			for (int j = 0; j < p->size(); ++j) 
				if (i != j && p->at(j)->getType() == B) {
					d.push_back(distE(p->at(i), p->at(j)));
				}

			sort(d.begin(), d.end());
			for (int j = 0; j < T; ++j) {
				acc+=d[j];
			}
		}
		if (i%1000 == 0) std::printf("acc = %e\n", acc);
	}

	return acc;
}

int main() {
	lane_maxwell::modules::matrix::matrixModuleInit(1e-7);

	Points* p = new Points;
	int nx, ny, nz, T;

	std::printf("(nx,ny,nz)=");
	std::scanf("%d%d%d", &nx, &ny, &nz);

	std::printf("T=");
	std::scanf("%d", &T);

	genPoints(p, nx, ny, nz, 'E');
	genPoints(p, nx, ny, nz, 'H');

	std::printf("Points...\n");
	//print(stdout, p);
	
	std::sort(p->begin(), p->end(),PointComp());

	//printAll(stdout, p);

	buildSupportDomain(stdout, 'E', 'H', T, p, 0.01, distE, 1);

	std::printf("r acc dumb = %e\n", dumb(p, T, 'E', 'H'));
	std::printf("r acc line sweep = %e\n", accSD(p));
	
	free(p);
	return 0;
}

File: octave/test_rpim_2.m
#GoGoRPIM - A RPIM implementation and 2-D Eletromagnetic simulator
#    Copyright (C) 2012  Péricles Lopes Machado (LANE-UFPA)
#
#    This program is free software: you can redistribute it and/or modify
#    it under the terms of the GNU General Public License as published by
#    the Free Software Foundation, either version 3 of the License, or
#    (at your option) any later version.
#
#    This program is distributed in the hope that it will be useful,
#    but WITHOUT ANY WARRANTY; without even the implied warranty of
#    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#    GNU General Public License for more details.
#
#    You should have received a copy of the GNU General Public License
#    along with this program.  If not, see <http://www.gnu.org/licenses/>.
#

function test_rpim_2(tipo)
    fmax = 1e9;
    lmb = 3e8 / fmax;
    dx = dy = lmb / 20;

    if tipo == 1
        f = fopen("pontos_igual.dat");
        fp = fopen("dist_1.dat","w+");
    end

    if tipo == 2
        f = fopen("pontos_igual.dat");
        fp = fopen("dist_2.dat","w+");    
    end

    if tipo == 3
        f = fopen("pontos_igual.dat");
        fp = fopen("dist_3.dat","w+");
    end

    m = 1;
    NR = 3;

    while (!feof(f))    
        [px, py, count] = fscanf(f,"%lf%lf", '2');

        if count == 0
            break;
        end

        xo(m) = x(m) = px * dx;
        yo(m) = y(m) = py * dy;

        
        if tipo == 3 && m > 1       
            x(m) = (1 + 0.2 * rand() - 0.1) * x(m);
            y(m) = (1 + 0.2 * rand() - 0.1) * y(m); 
        end

        if tipo == 2 && m > 1  && NR > 0     
            x(m) = (1 + 0.2 * rand() - 0.1) * x(m);
            y(m) = (1 + 0.2 * rand() - 0.1) * y(m);
            NR = NR - 1; 
        end


        fprintf(fp,"%e %e\n", x(m), y(m));       

        m = m+1;

    end

    fclose(fp);
    fclose(f);


    pts = [x' y'];

    N = size(pts)(1);
    M = size(pts)(2);
    pi = acos(-1);

    K = 2 * pi * fmax / 3e8;


    Fr = 1;


    dFr(1) = K; 
    dFr(2) = 0;

    

    errmax = 1e-10;
    NP = 24;

    cmin = 1;
    cmax = 101;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
    if tipo == 3
        M = 7;
    else
        M = 1;
    end
    
    


    if tipo == 1
        f11 = fopen("err_1lsfcm_1_1.dat","w+");
        f12 = fopen("err_1lsfcm_2_1.dat","w+");

        f21 = fopen("err_2lsfcm_1_1.dat","w+");
        f22 = fopen("err_2lsfcm_2_1.dat","w+");

        f31 = fopen("err_3lsfcm_1_1.dat","w+");
        f32 = fopen("err_3lsfcm_2_1.dat","w+");
    end

    if tipo == 2
        f11 = fopen("err_1lsfcm_1_2.dat","w+");
        f12 = fopen("err_1lsfcm_2_2.dat","w+");

        f21 = fopen("err_2lsfcm_1_2.dat","w+");
        f22 = fopen("err_2lsfcm_2_2.dat","w+");

        f31 = fopen("err_3lsfcm_1_2.dat","w+");
        f32 = fopen("err_3lsfcm_2_2.dat","w+");
    end

    if tipo == 3
        f11 = fopen("err_1lsfcm_1_3.dat","w+");
        f12 = fopen("err_1lsfcm_2_3.dat","w+");

        f21 = fopen("err_2lsfcm_1_3.dat","w+");
        f22 = fopen("err_2lsfcm_2_3.dat","w+");

        f31 = fopen("err_3lsfcm_1_3.dat","w+");
        f32 = fopen("err_3lsfcm_2_3.dat","w+");
    end

    printf("lsfcm\n");

    for NP = 2:2:16
        errm(1, NP) = 0;
        errm2(1, NP) = 0;

        errm(2, NP) = 0;
        errm2(2, NP) = 0;

        errm(3, NP) = 0;
        errm2(3, NP) = 0;
    end

    for NI = 1:M

        if tipo == 3
            for m = 2:N       
                x(m) = xo(m)+(0.02 * rand() - 0.01) * dx;
                y(m) = yo(m)+(0.02 * rand() - 0.01) * dy; 
            end
        end



        pts = [x' y'];

        F = [];
        dF = [];

        for n = 2:N
            F(n-1) = 0;
            F(n-1) = F(n-1) + sin(K*pts(n,1)) + cos(K*pts(n, 2));        
        end

        for  NP = 2:2:16

            printf("NP = %d\n", NP);

            cmin = 1;
            cmax = 10;
            [c, err, ni] = lsfcm(pts, fmax, cmin, cmax, NP, errmax);

            %[phi, dphi] = rpim(pts, c(1));
            %Fi = irpim(phi, F);
            %err(1) = Fi - Fr;

            
            %[phi, dphi] = rpim(pts, c(2));
            %Fi = irpim(dphi(1,:), F);
            %err(2) = Fi - dFr(1);

            
            %[phi, dphi] = rpim(pts, c(3));
            %Fi = irpim(dphi(2,:), F);
            %err(3) = Fi - dFr(2);

            errm(1, NP) = errm(1, NP) + abs(err(1));
            errm(2, NP) = errm(2, NP) + abs(err(2));
            errm(3, NP) = errm(3, NP) + abs(err(3));

            cmin = 1;
            cmax = 10;
            [c, err, ni] = lsfcm2(pts, fmax, cmin, cmax, NP, errmax);
           
            %[phi, dphi] = rpim(pts, c(1));
            %Fi = irpim(phi, F);
            %err(1) = Fi - Fr;

            
            %[phi, dphi] = rpim(pts, c(2));
            %Fi = irpim(dphi(1,:), F);
            %err(2) = Fi - dFr(1);

            
            %[phi, dphi] = rpim(pts, c(3));
            %Fi = irpim(dphi(2,:), F);
            %err(3) = Fi - dFr(2);

            errm2(1, NP) = errm2(1, NP) + abs(err(1));
            errm2(2, NP) = errm2(2, NP) + abs(err(2));
            errm2(3, NP) = errm2(3, NP) + abs(err(3));

        end

         printf("M = %d\n", M);

    end

    for NP = 2:2:16
        fprintf(f11, "%d %e\n", NP, errm(1, NP)/M);
        fprintf(f12, "%d %e\n", NP, errm2(1, NP)/M);

        fprintf(f21, "%d %e\n", NP, errm(2, NP)/M);
        fprintf(f22, "%d %e\n", NP, errm2(2, NP)/M);

        fprintf(f31, "%d %e\n", NP, errm(3, NP)/M);
        fprintf(f32, "%d %e\n", NP, errm2(3, NP)/M);
    end

    fclose(f11); fclose(f12);
    fclose(f21); fclose(f22);
    fclose(f31); fclose(f32);
end


